src/uct/tcp/tcp_ep.c

```diff
@@ -386,10 +386,11 @@ uct_tcp_ep_create_socket_and_connect(uct_tcp_iface_t *iface,
         uct_tcp_ep_destroy_internal(&ep->super.super);
     }
 err_close_fd:
-    /* fd has to be valid in case of valid EP has been
-     * passed to this function */
-    ucs_assert((*ep_p == NULL) || (fd != -1));
-    ucs_close_fd(&fd);
+    if (ep != NULL) {
+        ucs_close_fd(&ep->fd);
+    } else {
+        ucs_close_fd(&fd);
```

Comment:
are we closing fd which was not created in this function?

Reply:
no, this fd is created here, but it could be assigned to EP or not. If was assigned, i.e. EP isn't `NULL`, we have to close it from EP to fill by `-1`

Reply:
ok so why we can't always ucs_close_fd(&fd); ?

Reply:
> ok so why we can't always ucs_close_fd(&fd); ?

to not close then same FD twice from EP there:
https://github.com/openucx/ucx/blob/7f9a7f7a0d67a6143e163c378fbf76f47c15115f/src/uct/tcp/tcp_ep.c#L579

the `ucs_close_fd()` will set -1 to  the EP's fd

Reply:
looks weird
maybe move line 392 to line 363?

Reply:
> maybe move line 392 to line 363?

yes, it works
done